Remove legacy branch listing api

mtsgrd opened the pull request:
- renamed to `find_git_branches` by name
- was only used to get branches for preview
- was unnecessarily being reloaded constantly

<!-- GitButler Footer Boundary Top -->
---
This is **part 2 of 2 in a stack** made with GitButler:
- <kbd>&nbsp;2&nbsp;</kbd> #5573 👈 
- <kbd>&nbsp;1&nbsp;</kbd> #5565 
<!-- GitButler Footer Boundary Bottom -->

mtsgrd committed:
Remove legacy branch listing api

- renamed to `find_git_branches` by name
- was only used to get branches for preview
- was unnecessarily being reloaded constantly

ndom91 commented:
What do you think about using `<kbd>` elements? Combined some `emsp` we can even use the finger on the left and have it look alright imo

<!-- GitButler Footer Boundary Top -->
---
This is **part 2 of 2 in a stack** made with GitButler:
* <kbd>&nbsp;👉&nbsp;</kbd> #5573  
* <kbd>&emsp;1&emsp;</kbd> #5565 
<!-- GitButler Footer Boundary Bottom -->

Caleb-T-Owens commented on apps/desktop/src/routes/[projectId]/branch/[...name]/+page.svelte:
```diff
@@ -1,65 +1,36 @@
+{/if}
+{#each remoteBranches as remoteBranch}
+	<BranchPreview {remoteBranch} {pr} />
+{/each}
```
Is there a particular design wise reason we're changing thew way we display Local and Remote previews?

mtsgrd replied:
Right this is perhaps a mistake on my part. Can you confirm we do _not_ want to show the local branch if one or more remote branches exist?

mtsgrd replied:
I'll add back a loading state as well.

Caleb-T-Owens replied:
> Can you confirm we do not want to show the local branch if one or more remote branches exist?
Yes. If both a local and remote branch is provided to the component, it displays the local branch commits as the branch commits, and the remote commits as upstream ones; like how it is in the workspace

Caleb-T-Owens commented on crates/gitbutler-branch-actions/src/remote.rs:
```diff
@@ -60,32 +62,50 @@ pub struct RemoteCommit {
-        .flatten()
+    let repo = ctx.repository();
+    let mut branches = vec![];
+    let remotes_raw = repo.remotes()?;
+    let remotes = remotes_raw.iter().flatten().collect_vec();
```
Why do we need to flatten the remotes array? Isn't the return type `Result<Vec<String>>`?

mtsgrd replied:
We need to flatten it since remotes_raw is a git2::StringArray, which converts into `Vec<Option<&str>>`.

Caleb-T-Owens replied:
I see

Caleb-T-Owens commented on crates/gitbutler-branch-actions/src/remote.rs:
```diff
@@ -60,32 +62,50 @@ pub struct RemoteCommit {
+        .iter()
+        .filter_map(|remote| {
+            let branch_name = format!("{}/{}", remote, branch_name);
+            repo.find_branch(&branch_name, BranchType::Remote)
+                .ok()
+                .and_then(|branch| branch_to_remote_branch(&branch, &[remote]).ok())
+                .unwrap_or(None)
```
If I'm not mistaken, could this be simplified to:
```rs
            repo.find_branch(&branch_name, BranchType::Remote)
                 .and_then(|branch| branch_to_remote_branch(&branch, &[remote]))
                 .ok()
```

mtsgrd replied:
It can't, because that inner function returns a result of an option. But like I say in a comment below, let's change the inner function error on the conversion failing. It does not appear to me like there should be a legitimate reason for it to fail?

Caleb-T-Owens commented on crates/gitbutler-branch-actions/src/remote.rs:
```diff
@@ -60,32 +62,50 @@ pub struct RemoteCommit {
+                .map_err(|err| {
+                    tracing::warn!(?err, "Ignoring branch");
+                })
+                .ok()
+        })
+        .unwrap_or(None)
     {
-        let branch = match branch_to_remote_branch(&branch, &remotes) {
-            Ok(Some(b)) => b,
-            Ok(None) => continue,
-            Err(err) => {
-                tracing::warn!(?err, "Ignoring branch");
-                continue;
-            }
-        };
+        all_branches.push(local_branch);
+    }
```
Could we write this one as
```rs
// Let the compiler know we're swallowing the error
let _ = repo
         .find_branch(branch_name, BranchType::Local)
         .and_then(|branch| {
             let branch = branch_to_remote_branch(&branch, &remotes)?;
             all_branches.push(branch);
             Ok(())
         });
```

mtsgrd replied:
The question mark does not work inside of `.and_then(..)`. Also, `branch_to_remote_branch` returns an `Option`, but we could perhaps change that to an error.

Caleb-T-Owens replied:
The question mark does work inside of `and_then`. If you've got a function or closure whose return type is `Option`, then you can use `?` on any functions which return `Option` inside of it, and similar for `Result`.


I don't see any particularly good reason why branch_to_remote_branch should be `Result<Option<>>` as it a failure to convert. If it returns just a `Result` then my snippet should work.

Caleb-T-Owens commented on crates/gitbutler-branch-actions/src/remote.rs:
```diff
@@ -95,10 +115,10 @@ pub fn list_local_branches(ctx: &CommandContext) -> Result<Vec<RemoteBranch>> {
             && branch.name.branch() != Some("gitbutler/target")
         {
-            remote_branches.push(branch);
+            branches.push(branch.to_owned());
```
Could we make sure that the `branches` array is filled with the owned types already so we don't need to clone them?

Caleb-T-Owens commented on crates/gitbutler-branch-actions/src/virtual.rs:
```diff
@@ -455,7 +456,8 @@ pub fn list_virtual_branches_cached(
         let upstream = upstream_branch.and_then(|upstream_branch| {
-            let remotes = repo.remotes().ok()?;
+            let remotes_raw = repo.remotes().ok()?;
+            let remotes = remotes_raw.into_iter().flatten().collect_vec();
```
@Byron do you have an opinion on itertools? In recent weeks I've seen it get used more and more, and none of the uses seem to be much of a gain over doing it without.

mtsgrd replied:
The alternative feels a bit more verbose? Either:
```
let remotes: Vec<_> = remotes_raw.into_iter().flatten().collect();
```
or
```
let remotes = remotes_raw.into_iter().flatten().collect::<Vec<_>>();
```

mtsgrd replied:
I think if they are functionally I don't mind using `.collect()`, but which of the two styles I mention above do you guys prefer?

Caleb-T-Owens replied:
collect_vec() is the itertools, so the plain rust collect() is the preferable one.

mtsgrd replied:
Yes but it often requires explicit types. Just wondering if there's a better option between putting it at the beginning of the line, or next to the collect_vec call, as mentioned in my previous comment.

Caleb-T-Owens replied:
I think `let remotes = remotes_raw.into_iter().flatten().collect::<Vec<_>>();` is preferable

Byron replied:
Sometimes the `Vec<_>` can be inferred, but probably not in this situation.

To my mind, using `itertools` here is fine as it is intuitive and thus is fast and easy to pull off. And that matters.
For `gitoxide` I can't use it, as I get angry letters sent to my house when a new dependency shows up downstream, and it could hide allocations which I rather keep explicit (e.g. `it.sorted()`).

But in application code I don't think there is a good enough reason not to use it.

Caleb-T-Owens commented on crates/gitbutler-cli/src/main.rs:
```diff
@@ -35,7 +35,7 @@ fn main() -> Result<()> {
                     short_tracking_branch_name,
                 }) => command::vbranch::set_base(project, short_tracking_branch_name),
-                Some(vbranch::SubCommands::ListLocal) => command::vbranch::list_local(project),
+                Some(vbranch::SubCommands::ListLocal) => command::vbranch::list_all(project),
```
It would be reasonable to deem this out of scope, but we should add a TODO here because the command name is clearly wrong

mtsgrd replied:
Good point. I need to change this line since I remove `list_local` function. I'll update the enum to `SubCommand:List`, and update the comment on the definition.

mtsgrd committed:
Rename RemoteBranchService -> GitBranchService

mtsgrd committed:
Rename and move remoteBranches.ts -> gitBranch.ts

mtsgrd committed:
Rename remoteBranchService variables -> gitBranchService

mtsgrd commented on crates/gitbutler-branch-actions/src/remote.rs:
```diff
@@ -60,45 +61,51 @@ pub struct RemoteCommit {
 /// a list of all the normal (non-gitbutler) git branches.
-pub fn list_local_branches(ctx: &CommandContext) -> Result<Vec<RemoteBranch>> {
-    let default_target = default_target(&ctx.project().gb_dir())?;
+pub fn find_git_branches(ctx: &CommandContext, branch_name: &str) -> Result<Vec<RemoteBranch>> {
+    let repo = ctx.repository();
+    let remotes_raw = repo.remotes()?;
+    let remotes: Vec<_> = remotes_raw.iter().flatten().collect();
+
+    // We since we are testing for the presence of branches we swallow any errors
+    // from both finding the branch, and looking up the branch commit data. The
+    // latter can fail if/when a ref points to a object that doesn't exist.
+    let mut all_branches: Vec<RemoteBranch> = remotes
+        .iter()
+        .filter_map(|remote_name| {
+            let branch_name = format!("{}/{}", remote_name, branch_name);
+            let blah = repo
+                .find_branch(&branch_name, BranchType::Remote)
+                .ok()
+                .and_then(|branch| branch_to_remote_branch(&branch, &[remote_name]).ok());
+            blah
+        })
+        .collect();
```
@Caleb-T-Owens I think it's best if the functions return errors, since one would expect them to work if you are looking up a valid repo and branch. In the case of `get_remote_branch_data` it's natural to fail the front-end request, while here in a "find" function it feels ok to swallow them?

Caleb-T-Owens replied:
Can we talk about this on a call tomorrow?

mtsgrd commented on crates/gitbutler-branch-actions/src/remote.rs:
```diff
@@ -60,45 +61,51 @@ pub struct RemoteCommit {
+    let target_branch = &default_target(&ctx.project().gb_dir())?.branch;
+    Ok(all_branches
+        .into_iter()
+        .filter(move |branch| {
```
The `move` isn't necessary, I'll remove it.

ndom91 commented:
> @ndom91 that misalignment introduced by the emoji hurts my eyes a bit. Could you suggest that in a different pr if you are motivated about it? :) 

Hahah yeah I was playing around with all the different space types to try to get it to line up perfectly, but no dice. 

If we drop the emoji in the front though I think just the digits, `kbd` and some `nbsp`'s would be nice, for example: 

- <kbd>&nbsp;1&nbsp;</kbd>👈 blah blah blah 
- <kbd>&nbsp;2&nbsp;</kbd> foo bar baz 

I can bring it up elsewhere 👍

Caleb-T-Owens reviewed (approved):
LGTM

mtsgrd committed:
Drop RemoteBranch in favor of RemoteBranchData

- redundant data structure and api removed
- served same purpose as `find_git_branches`
